File: python/creational/prototype.py
import copy

# prototype interface
class DocumentPrototype:
    def clone(self):
        return copy.deepcopy(self)

# concerete prototype implementation
class Invoive(DocumentPrototype):
    def __init__(self, item, price, amount):
        self.type = "Invoice"
        self.item = item
        self.price = price
        self.amount = amount
    
    def __str__(self):
        return f"{self.type} - Amount: ${self.amount}"
    
class Report(DocumentPrototype):
    def __init__(self, id, total_price):
        self.type = "Report"
        self.id = id
        self.total_price = total_price

    def __str__(self):
        return f"{self.type} - Id: {self.id}, Total Price: {self.total_price}"
    
if __name__ == "__main__":

    invoice = Invoive("default", 0, 0)
    report = Report(0, 0)

    new_invoice = invoice.clone()
    new_invoice.item = "PR02"
    new_invoice.amount = 25

    print(new_invoice)

    new_report = report.clone()
    new_report.id = 2
    new_report.total_price = 5000000

    print(new_report)



    
        
        

File: go/creational/prototype..go
package main

import "fmt"

// prototype interface
type DocumentPrototype interface {
	Clone() DocumentPrototype
}

// concerete prototype implementation
type Invoice struct {
	item  string
	price float64
	amout int
}

func (i *Invoice) Clone() DocumentPrototype {
	return &Invoice{
		item:  i.item,
		price: i.price,
		amout: i.amout,
	}
}

func (i *Invoice) String() string {
	return fmt.Sprintf("Invoice - Item: %s, Price: $%.2f, Amount: %d", i.price, i.price, i.amout)
}

type OrderReport struct {
	id          uint64
	totalAmount int
}

func (r *OrderReport) Clone() DocumentPrototype {
	return &OrderReport{
		id:          r.id,
		totalAmount: r.totalAmount,
	}
}

func (i *OrderReport) String() string {
	return fmt.Sprintf("Report - Id: %d, Total Amount: %d", i.id, i.totalAmount)
}

func main() {

	invoice := Invoice{}

	newInvoice := invoice.Clone().(*Invoice)
	newInvoice.item = "P02"
	newInvoice.amout = 5
	newInvoice.price = 2000.0

	fmt.Println(newInvoice)

	report := OrderReport{}

	newReport := report.Clone().(*OrderReport)
	newReport.id = 5
	newReport.totalAmount = 260

	fmt.Println(newReport)
}
